src/InodeCache.cpp

```diff
@@ -0,0 +1,528 @@
+const uint32_t k_num_buckets = 32 * 1024;
+const uint32_t k_num_entries = 4;
+
+static_assert(sizeof(digest::bytes) == 20,
+              "Increment version number if size of digest  is changed.");
+
+const char k_default_basename[] = "/inode_cache";
+
+struct Key
+{
+  dev_t st_dev;
+  ino_t st_ino;
+  mode_t st_mode;
+#  ifdef HAVE_STRUCT_STAT_ST_MTIM
+  timespec st_mtim;
+#  else
+  time_t st_mtim;
+#  endif
+#  ifdef HAVE_STRUCT_STAT_ST_CTIM
+  timespec st_ctim; // Included for sanity checking.
+#  else
+  time_t st_ctim; // Included for sanity checking.
+#  endif
+  off_t st_size; // Included for sanity checking.
+  bool sloppy_time_macros;
+};
+
+struct Entry
+{
+  digest key_digest;  // Hashed key
+  digest file_digest; // Cached file hash
+  int return_value;   // Cached return value
+};
+
+struct Bucket
+{
+  pthread_mutex_t mt;
+  int32_t hits;
+  int32_t misses;
+  Entry entries[k_num_entries];
+};
+
+struct SharedRegion
+{
+  uint32_t version;
+  std::atomic<int64_t> errors;
+  Bucket buckets[k_num_buckets];
+};
+
+SharedRegion* g_sr;
+
+bool
+mmap_file(const std::string& inode_cache_file)
+{
+  if (g_sr) {
+    munmap(g_sr, sizeof(SharedRegion));
+    g_sr = nullptr;
+  }
+  int fd = open(inode_cache_file.c_str(), O_RDWR);
+  if (fd < 0) {
+    cc_log("Failed to open inode cache %s: %s",
+           inode_cache_file.c_str(),
+           strerror(errno));
+    return false;
+  }
+  SharedRegion* sr = reinterpret_cast<SharedRegion*>(mmap(
+    nullptr, sizeof(SharedRegion), PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0));
+  close(fd);
+  if (sr == reinterpret_cast<void*>(-1)) {
+    fprintf(stderr,
+            "Failed to mmap %s: %s\n",
+            inode_cache_file.c_str(),
+            strerror(errno));
+    return false;
+  }
+  // Drop the file from disk if the found version is not matching. This will
+  // allow a new file to be generated.
+  if (sr->version != k_version) {
+    cc_log(
+      "Dropping inode cache because found version %u does not match expected "
+      "version %u",
+      sr->version,
+      k_version);
+    munmap(sr, sizeof(SharedRegion));
+    unlink(inode_cache_file.c_str());
+    return false;
+  }
+  g_sr = sr;
+  return true;
+}
+
+bool
+hash_inode(const Config& config, const char* path, digest* digest)
+{
+  struct stat st_buf;
+  if (stat(path, &st_buf)) {
+    cc_log("Could not stat %s: %s", path, strerror(errno));
+    return false;
+  }
+
+  Key key;
+  memset(&key, 0, sizeof(Key));
+  key.st_dev = st_buf.st_dev;
+  key.st_ino = st_buf.st_ino;
+  key.st_mode = st_buf.st_mode;
+#  ifdef HAVE_STRUCT_STAT_ST_MTIM
+  key.st_mtim = st_buf.st_mtim;
+#  else
+  key.st_mtim = st_buf.st_mtime;
+#  endif
+#  ifdef HAVE_STRUCT_STAT_ST_CTIM
+  key.st_ctim = st_buf.st_ctim;
+#  else
+  key.st_ctim = st_buf.st_ctime;
+#  endif
+  key.st_size = st_buf.st_size;
+  key.sloppy_time_macros = config.sloppiness() & SLOPPY_TIME_MACROS;
+
+  struct hash* hash = hash_init();
```

Comment:
Need to use blake2 as the digest is actually stored in the entries instead of the original key (as the suggest is shorter and we need to hash the key anyway). This is fine as long as we don't use a weaker hash function than we used to hash the source files.

Reply:
Right, not sure what I was thinking there.